src/rgw/rgw_auth_keystone.cc

```diff
@@ -0,0 +1,475 @@
+  ldout(cct, 20) << "token_id=" << token_id << dendl;
+
+  /* Check cache first. */
+  t = token_cache.find(token_id);
+  if (t) {
+    ldout(cct, 20) << "cached token.project.id=" << t->get_project_id()
+                   << dendl;
+    auto apl = apl_factory->create_apl_remote(cct, s, get_acl_strategy(*t),
+                                              get_creds_info(*t, roles.admin));
+    return result_t::grant(std::move(apl));
+  }
+
+  /* Retrieve token. */
+  if (rgw_is_pki_token(token)) {
+    try {
+      t = decode_pki_token(token);
+    } catch (...) {
+      /* Last resort. */
+      t = get_from_keystone(token);
+    }
+  } else {
+    /* Can't decode, just go to the Keystone server for validation. */
+    t = get_from_keystone(token);
+  }
+
+  if (! t) {
+    return result_t::deny(-EACCES);
+  }
+
+  /* Verify expiration. */
+  if (t->expired()) {
+    ldout(cct, 0) << "got expired token: " << t->get_project_name()
+                  << ":" << t->get_user_name()
+                  << " expired: " << t->get_expires() << dendl;
+    return result_t::deny();
+  }
+
+  /* Check for necessary roles. */
+  for (const auto& role : roles.plain) {
+    if (t->has_role(role) == true) {
+      ldout(cct, 0) << "validated token: " << t->get_project_name()
+                    << ":" << t->get_user_name()
+                    << " expires: " << t->get_expires() << dendl;
+      token_cache.add(token_id, *t);
+      auto apl = apl_factory->create_apl_remote(cct, s, get_acl_strategy(*t),
+                                            get_creds_info(*t, roles.admin));
+      return result_t::grant(std::move(apl));
+    }
+  }
+
+  ldout(cct, 0) << "user does not hold a matching role; required roles: "
+                << g_conf->rgw_keystone_accepted_roles << dendl;
+
+  return result_t::deny();
+}
+
+
+/*
+ * Try to validate S3 auth against keystone s3token interface
+ */
+std::pair<boost::optional<rgw::keystone::TokenEnvelope>, int>
+EC2Engine::get_from_keystone(const std::string& access_key_id,
+                             const std::string& string_to_sign,
+                             const std::string& signature) const
+{
+  /* prepare keystone url */
+  std::string keystone_url = config.get_endpoint_url();
+  if (keystone_url.empty()) {
+    throw -EINVAL;
+  }
+
+  const auto api_version = config.get_api_version();
+  if (config.get_api_version() == rgw::keystone::ApiVersion::VER_3) {
+    keystone_url.append("v3/s3tokens");
+  } else {
+    keystone_url.append("v2.0/s3tokens");
+  }
+
+  /* get authentication token for Keystone. */
+  std::string admin_token;
+  int ret = rgw::keystone::Service::get_admin_token(cct, token_cache, config,
+                                                    admin_token);
+  if (ret < 0) {
+    ldout(cct, 2) << "s3 keystone: cannot get token for keystone access"
+                  << dendl;
+    throw ret;
+  }
+
+  using RGWValidateKeystoneToken
+    = rgw::keystone::Service::RGWValidateKeystoneToken;
```

Comment:
This is much, much better! I've completely forgotten about this possibility. Thanks, Casey!